core/modules/replica/tools/qserv-replica-master.cc

```diff
@@ -35,715 +40,322 @@
-    std::vector<std::string> databaseFamilies;
-
-    /// The flag to indicate a catastrophic activity which must result
-    /// in a termination of the application
-    std::atomic<bool> failed;
-    
-    /// This flag is raised by the health monitoring thread to notify
-    /// the replication thread that it should wrap up its operation
-    /// and quit
-    std::atomic<bool> stopReplication;
+    /**
+     * The factory method is the only way of creating objects of this class
+     * because of the very base class's inheritance from 'enable_shared_from_this'.
+     *
+     * @param argc
+     *   the number of command-line arguments
+     *
+     * @param argv
+     *   the vector of command-line arguments
+     *
+     * @param description
+     *   the description of the application to be printed when the command-line
+     *   parser is invoked in the "help" mode.
+     *
+     * @param logger
+     *   the name of a logger
+     */
+    static Ptr create(int argc,
+                      const char* const argv[],
+                      std::string const& description,
+                      std::string const& logger) {
+        return Ptr(
+            new MasterControllerApp(
+                argc,
+                argv,
+                description,
+                logger
+            )
+        );
+    }
 
-    /// no parent for any job initited by the application
-    std::string const parentJobId;
+    // Default construction and copy semantics are prohibited
 
-    /// Force Qserv synchronization if 'true'
-    bool forceQservSync;
+    MasterControllerApp() = delete;
+    MasterControllerApp(MasterControllerApp const&) = delete;
+    MasterControllerApp& operator=(MasterControllerApp const&) = delete;
 
-    /// Permanently delete workers if set to 'true'
-    bool permanentDelete;
+    ~MasterControllerApp() override = default;
 
-    // Context strings for threads
+protected:
 
-    std::string healthMonitorContext;
-    std::string replicationLoopContext;
+    /**
+     * @see MasterControllerApp::create()
+     */
+    MasterControllerApp(int argc,
+                        const char* const argv[],
+                        std::string const& description,
+                        std::string const& logger)
+        :   Application(
+                argc,
+                argv,
+                description,
+                true /* injectDatabaseOptions */,
+                true /* boostProtobufVersionCheck */,
+                true /* enableServiceProvider */
+            ),
+            _healthProbeIntervalSec  (60),
+            _replicationIntervalSec  (60),
+            _workerResponseTimeoutSec(60),
+            _workerEvictTimeoutSec   (3600),
+            _qservSyncTimeoutSec     (60),
+            _numReplicas             (0),
+            _numIter                 (0),
+            _purge          (false),
+            _forceQservSync (false),
+            _permanentDelete(false),
+            _failed         (false),
```

Comment:
would it be better to define these defaults in a cfg or with defines at the top of the file or something?

Reply:
Yes, I was already considering moving these parameters into the Configuration service (and eventually - into the Configuration database).